pandas/core/internals/blocks.py

```diff
@@ -434,40 +434,56 @@ def fillna(
+        arr = self if inplace else self.copy()
+        limit = libalgos.validate_limit(
+            len(self) if self.ndim == 1 else self.shape[1], limit=limit
+        )
 
         if not self._can_hold_na:
-            if inplace:
-                return [self]
+            return [arr]
+
+        if not self.is_extension:
+            if self._can_hold_element(value):
+                if self.ndim == 1:
+                    if is_list_like(value):
+                        # TODO: Verify EA case
+                        if is_extension_array_dtype(value):
+                            mask = value.isna()
+                            value = np.asarray(value[mask], dtype=object)
+                            libalgos.fillna1d_multi_values(
+                                arr.values[mask], value=value, limit=limit
+                            )
+                        else:
+                            libalgos.fillna1d_multi_values(
+                                arr.values, value=value, limit=limit
+                            )
+                    else:
+                        libalgos.fillna1d(arr.values, value=value, limit=limit)
+                else:
+                    libalgos.fillna2d(arr.values, value=value, limit=limit)
+                return arr._maybe_downcast([arr], downcast)
+            elif self.ndim == 1 or self.shape[0] == 1:
+                blk = self.coerce_to_target_dtype(value)
+                # bc we have already cast, inplace=True may avoid an extra copy
+                return blk.fillna(value, limit=limit, inplace=True, downcast=None)
             else:
-                return [self.copy()]
-
-        if self._can_hold_element(value):
-            nb = self if inplace else self.copy()
-            putmask_inplace(nb.values, mask, value)
-            return nb._maybe_downcast([nb], downcast)
-
-        if noop:
-            # we can't process the value, but nothing to do
-            return [self] if inplace else [self.copy()]
-
-        elif self.ndim == 1 or self.shape[0] == 1:
-            blk = self.coerce_to_target_dtype(value)
-            # bc we have already cast, inplace=True may avoid an extra copy
-            return blk.fillna(value, limit=limit, inplace=True, downcast=None)
-
+                # operate column-by-column
+                return self.split_and_operate(
+                    type(self).fillna,
+                    value,
+                    limit=limit,
+                    inplace=inplace,
+                    downcast=None,
+                )
         else:
-            # operate column-by-column
-            return self.split_and_operate(
-                type(self).fillna, value, limit=limit, inplace=inplace, downcast=None
-            )
+            # TODO: This seems to work for EAS, verify it does
+            return [
+                self.make_block_same_class(
+                    values=self.values.fillna(value=value, limit=limit)
```

Comment:
you'll probably need to make NDArrayBackedExtensionArray.fillna use the new cython function(s)